src/html/data-sections.js

```diff
@@ -0,0 +1,156 @@
+    if (node.value && pattern.test(node.value)) {
+      handlefn(node, 'value');
+    }
+    if (node.alt && pattern.test(node.alt)) {
+      handlefn(node, 'alt');
+    }
+    if (node.url && pattern.test(node.url)) {
+      handlefn(node, 'url');
+    }
+    if (node.title && pattern.test(node.title)) {
+      handlefn(node, 'title');
+    }
+  });
+}
+
+function hasPlaceholders(section) {
+  try {
+    findPlaceholders(section, () => {
+      throw new Error('Placeholder detected');
+    });
+    return false;
+  } catch {
+    return true;
+  }
+}
+
+function fillPlaceholders(section) {
+  if (!section.meta.embedData && !Array.isArray(section.meta.embedData)) {
+    return;
+  }
+  const data = section.meta.embedData;
+  // required to make deepclone below work
+  removePosition(section);
+
+  const children = data.reduce((p, value) => {
+    const workingcopy = deepclone(section);
+
+    findPlaceholders(workingcopy, (node, prop) => {
+      if (typeof node[prop] === 'string') {
+        node[prop] = node[prop].replace(pattern, (_, expr) => value[expr]);
+      }
+    });
+    return [...p, ...workingcopy.children];
+  }, []);
+
+  section.children = children;
+}
+
+function normalizeLists(section) {
+  function cleanupLists([first, second]) {
+    // two consecutive, identical lists
+    if (first.type === 'list' && second.type === 'list'
+        && first.ordered === second.ordered
+        && first.start === second.start
+        && first.spread === second.spread) {
+      // move the children of the first to the second list
+      second.children = [...first.children, ...second.children];
+      first.children = [];
+      // mark the first list to be emptied
+      return first;
+    }
+    return null;
+  }
+
+  // get the sequence of lists to be removed
+  const emptiedLists = pipe(
+    trySlidingWindow(section.children, 2),
+    map(cleanupLists),
+    reject(is(null)),
+    list,
+  );
+
+  // perform the cleanup
+  section.children = pipe(
+    section.children, // take all existing children
+    reject((child) => contains(is(child))(emptiedLists)), // reject if they are in the list above
+    list, // make a nice array because the rest of the world doesn't like iterators yet
+  );
+}
+
+async function fillDataSections({ content: { mdast } }, { downloader }) {
+  async function extractData(section) {
+    return pmap(section, async (node) => {
+      if (node.type === 'dataEmbed') {
+        const task = downloader.getTaskById(`dataEmbed:${node.url}`);
+        const downloadeddata = await task;
+        // TODO: better error handling
+        // TODO: check that the result is an array
+        const json = await downloadeddata.json();
+        section.meta.embedData = json;
+      }
+      return node;
+    });
+  }
+
+  const dataSections = selectAll('section', mdast);
+
+  // extract data from all sections
+  await Promise.all(dataSections.map(extractData));
+
+  if (dataSections.length === 0 && hasPlaceholders(mdast)) {
```

Comment:
I fear, that this will be very expensive, especially for larger trees... ideally, you can just walk the entire tree once and process it...

Reply:
Makes sense, intuitively. In 1e0e1d1 I've added two tests with larger Markdown documents – one without data embeds, one with a small data embed. Each test will process the same file 10 times, which leads to following results:

<img width="666" alt="Screen Shot 2020-03-11 at 09 52 39" src="https://user-images.githubusercontent.com/39613/76399274-a1cfca00-637e-11ea-863f-72f26ef42b31.png">

The checked case is faster than the non-checked case, by about one standard deviation.

Reply:
but isn't it more efficient to check if the section/document has a `dataEmbed` first, before looking for placeholders?

Reply:
ok.